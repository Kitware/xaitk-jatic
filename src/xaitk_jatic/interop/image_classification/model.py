"""
This module provides the `JATICImageClassifier` class, an adapter for integrating JATIC-based
image classifiers with the SMQTK `ClassifyImage` interface. It enables the use of JATIC
protocol-based classifiers in pipelines that require the SMQTK interface.

Classes:
    JATICImageClassifier: Adapts a JATIC image classification model for compatibility with
                          SMQTK's `ClassifyImage` interface.

Dependencies:
    - maite.protocols.image_classification: For the JATIC image classification protocol.
    - numpy: For numerical operations.
    - smqtk_classifier: For SMQTK classifier interfaces and classification utilities.
"""

from collections.abc import Hashable, Iterator, Sequence

import maite.protocols.image_classification as ic
import numpy as np
from smqtk_classifier.interfaces.classification_element import CLASSIFICATION_DICT_T
from smqtk_classifier.interfaces.classify_image import IMAGE_ITER_T, ClassifyImage
from typing_extensions import override


class JATICImageClassifier(ClassifyImage):
    """
    Adapter for the JATIC image classification protocol, implementing the SMQTK `ClassifyImage` interface.

    This adapter allows a JATIC protocol-based classifier to be used in SMQTK pipelines by transforming
    classification outputs into the expected format.

    Attributes:
        _classifier (ic.Model): The JATIC protocol-based image classifier instance.
        _ids (dict[int, Hashable]): A dictionary mapping label IDs to human-readable names.
        _img_batch_size (int): The number of images to process in a single batch.

    Methods:
        get_labels: Retrieve the human-readable class labels.
        classify_images: Perform classification on a batch of input images.
        get_config: Raise a `NotImplementedError` for serialization (not yet supported).
    """

    def __init__(
        self,
        classifier: ic.Model,
        ids: Sequence[int],
        img_batch_size: int = 1,
    ) -> None:
        """
        Initialize the JATICImageClassifier with a JATIC protocol-based classifier.

        Args:
            classifier (ic.Model): The JATIC protocol-based image classification model.
            ids (dict[int, Hashable]): A dictionary mapping label IDs to human-readable names.
            img_batch_size (int, optional): The number of images to process in a single batch. Defaults to 1.
        """
        self._classifier = classifier
        self._ids = sorted(ids)
        self._img_batch_size = img_batch_size

    @override
    def get_labels(self) -> Sequence[Hashable]:
        return self._ids

    @override
    def classify_images(self, img_iter: IMAGE_ITER_T) -> Iterator[CLASSIFICATION_DICT_T]:  # noqa: C901
        all_out = list()
        batch = list()

        # Convert from channels last to channels first
        # Channels first is specified in protocols v0.5.0
        def _to_channels_first(img: np.ndarray) -> np.ndarray:
            if img.ndim < 3:
                return img
            return np.moveaxis(img, -1, 0)

        # Transform outputs of a single batch
        def _generate_outputs(batch: Sequence[np.ndarray]) -> list[dict[int, float]]:
            predictions = np.asarray(self._classifier(batch))

            return [dict(enumerate(pred)) for pred in predictions]

        # Batch model passes
        for img in img_iter:
            batch.append(_to_channels_first(img))

            if len(batch) == self._img_batch_size:
                all_out.extend(_generate_outputs(batch))
                batch = list()

        # Leftover batch
        if len(batch) > 0:
            _generate_outputs(batch)

        return iter(all_out)

    @override
    def get_config(self) -> dict:
        raise NotImplementedError(
            "Constructor arguments are not serializable as is and require further implementation to do so.",
        )
