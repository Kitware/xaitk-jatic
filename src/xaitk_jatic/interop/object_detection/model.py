from collections.abc import Hashable, Iterable, Sequence
from typing import Dict, Tuple

import maite.protocols.object_detection as od
import numpy as np
from smqtk_detection import DetectImageObjects
from smqtk_image_io import AxisAlignedBoundingBox


class JATICDetector(DetectImageObjects):
    """Adapter for JATIC object detection protocol.

    :param detector: The JATIC protocol-based detector.
    :param id_to_name: Mapping from label IDs to names.
    :param img_batch_size: Image batch size for inference.
    """

    def __init__(
        self,
        detector: od.Model,
        ids: Sequence[int],
        img_batch_size: int = 1,
    ):
        self._detector = detector
        self._ids = sorted(ids)
        self._img_batch_size = img_batch_size

    def detect_objects(
        self,
        img_iter: Iterable[np.ndarray],
    ) -> Iterable[Iterable[Tuple[AxisAlignedBoundingBox, Dict[Hashable, float]]]]:
        all_out = list()

        # Convert from channels last to channels first
        # Channels first is specified in protocols v0.5.0
        def _to_channels_first(img: np.ndarray) -> np.ndarray:
            if img.ndim < 3:
                return img
            return np.moveaxis(img, -1, 0)

        # Combine JATIC detections for same bbox into one DetectImageObject detection
        def _xform_dets(
            bboxes: Iterable[AxisAlignedBoundingBox],
            labels: np.ndarray,
            probs: np.ndarray,
        ) -> Iterable[Tuple[AxisAlignedBoundingBox, Dict[Hashable, float]]]:
            dets_dict: Dict[AxisAlignedBoundingBox, Dict[Hashable, float]] = dict()
            for box, label, prob in zip(bboxes, labels, probs):
                if box not in dets_dict:
                    dets_dict[box] = {la: 0.0 for la in self._ids}
                dets_dict[box][label] = prob

            return [(box, prob_dict) for box, prob_dict in dets_dict.items()]

        # Convert bboxes into AxisAlignedBoundingBoxes
        # Protocols specify X0, Y0, X1, Y1 format in v0.5.0
        def _xform_bbox(box: np.ndarray) -> AxisAlignedBoundingBox:
            return AxisAlignedBoundingBox(box[0:2], box[2:4])

        # Convert protocol-based detector output into DetectImageObject format
        def _generate_outputs(batch: Sequence[np.ndarray]) -> None:
            predictions = self._detector(batch)

            for pred in predictions:
                boxes = [_xform_bbox(box) for box in np.asarray(pred.boxes)]
                labels = np.asarray(pred.labels)
                scores = np.asarray(pred.scores)

                all_out.append(_xform_dets(bboxes=boxes, labels=labels, probs=scores))

        # Batch model passes
        batch = list()
        for img in img_iter:
            batch.append(_to_channels_first(img))

            if len(batch) == self._img_batch_size:
                _generate_outputs(batch)
                batch = list()

        # Leftover batch
        if len(batch) > 0:
            _generate_outputs(batch)

        return all_out

    def get_config(self) -> dict:
        raise NotImplementedError(
            "Constructor arguments are not serializable as is and require further implementation to do so.",
        )
