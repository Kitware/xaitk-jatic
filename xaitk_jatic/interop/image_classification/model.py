from typing import Dict, Hashable, Iterator, Sequence
import numpy as np

from smqtk_classifier.interfaces.classify_image import ClassifyImage, IMAGE_ITER_T
from smqtk_classifier.interfaces.classification_element import CLASSIFICATION_DICT_T

import maite.protocols.image_classification as ic


class JATICImageClassifier(ClassifyImage):
    """
    Adapter for JATIC classifier protocol.

    :param classifier: The JATIC protocol-based classifier.
    :param id2name: Mapping from label IDs to names.
    :param img_batch_size: Image batch size for inference.
    """

    def __init__(
        self,
        classifier: ic.Model,
        id2name: Dict[int, Hashable],
        img_batch_size: int = 1
    ):
        self._classifier = classifier
        self._id2name = dict(sorted(id2name.items()))
        self._img_batch_size = img_batch_size

    def get_labels(self) -> Sequence[Hashable]:
        return [self._id2name[id] for id in sorted(self._id2name.keys())]

    def classify_images(
      self,
      img_iter: IMAGE_ITER_T
    ) -> Iterator[CLASSIFICATION_DICT_T]:
        all_out = list()
        batch = list()

        # Convert from channels last to channels first
        # Channels first is specified in protocols v0.5.0
        def _to_channels_first(img: np.ndarray) -> np.ndarray:
            if img.ndim < 3:
                return img
            return np.moveaxis(img, -1, 0)

        # Transform outputs of a single batch
        def _generate_outputs(batch: Sequence[np.ndarray]) -> None:
            predictions = np.asarray(self._classifier(np.asarray(batch)))

            for pred in predictions:
                all_out.append({self._id2name[id]: score for id, score in enumerate(pred)})

        # Batch model passes
        for img in img_iter:
            batch.append(_to_channels_first(img))

            if len(batch) == self._img_batch_size:
                _generate_outputs(batch)
                batch = list()

        # Leftover batch
        if len(batch) > 0:
            _generate_outputs(batch)

        return iter(all_out)

    def get_config(self) -> dict:
        raise NotImplementedError(
            "Constructor arguments are not serializable as is and require"
            "further implementation to do so."
        )
