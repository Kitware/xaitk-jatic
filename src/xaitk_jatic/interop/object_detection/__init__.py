"""Jatic compliant dataset and model objects for using XAITK for object detection"""
