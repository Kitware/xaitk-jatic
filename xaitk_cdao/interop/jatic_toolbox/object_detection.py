from typing import Iterable, Tuple, Dict, Hashable
import numpy as np

from smqtk_detection import DetectImageObjects
from smqtk_image_io import AxisAlignedBoundingBox

from jatic_toolbox.protocols import ObjectDetector


class JATICDetector(DetectImageObjects):
    """
    Wrapper for JATIC object detection protocol.
    """

    def __init__(self, detector: ObjectDetector):
        self._detector = detector

    def detect_objects(
      self,
      img_iter: Iterable[np.ndarray]
    ) -> Iterable[Iterable[Tuple[AxisAlignedBoundingBox, Dict[Hashable, float]]]]:
        detector_output = self._detector(img_iter)

        boxes = detector_output.boxes
        labels = detector_output.labels
        scores = detector_output.scores
        for boxes_per_image, labels_per_image, scores_per_image in zip(boxes, labels, scores):
            dets_dict = {}  # type: Dict[Tuple, Dict[Hashable, float]]
            for box, label, score in zip(boxes_per_image, labels_per_image, scores_per_image):
                b = tuple(box)
                if b not in dets_dict:
                    dets_dict[b] = {}
                dets_dict[b][label] = score

            dets = [(AxisAlignedBoundingBox(box[0:2], box[2:4]), score_dict) for box, score_dict in dets_dict.items()]
            yield dets

    def get_config(self) -> dict:
        return {
            "detector": self._detector
        }
