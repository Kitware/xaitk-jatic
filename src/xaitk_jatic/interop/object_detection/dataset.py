import copy
import logging
from dataclasses import dataclass
from typing import Any, Dict, Sequence, Tuple

import numpy as np
from maite.protocols.object_detection import (
    Dataset,
    DatumMetadataType,
    InputType,
    TargetType,
)
from PIL import Image  # type: ignore

try:
    import kwcoco  # type: ignore

    is_usable = True
except ImportError:
    is_usable = False

OBJ_DETECTION_DATUM_T = Tuple[InputType, TargetType, DatumMetadataType]

LOG = logging.getLogger(__name__)


@dataclass
class JATICDetectionTarget:
    """Dataclass for the datum-level JATIC output detection format."""

    boxes: np.ndarray
    labels: np.ndarray
    scores: np.ndarray


if not is_usable:
    LOG.warning(
        "COCOJATICObjectDetectionDataset requires additional dependencies, please install 'xaitk-jatic[tools]'."
    )
else:

    class COCOJATICObjectDetectionDataset(Dataset):
        """Dataset class to convert a COCO dataset to a dataset compliant with JATIC's Object Detection protocol.

        Parameters
        ----------
        kwcoco_dataset : kwcoco.CocoDataset
            The kwcoco COCODataset object.
        image_metadata : Dict[int, Dict[str, Any]]
            A dict of per-image metadata, by image id. Any metadata required by a perturber should be provided.
        skip_no_anns: bool
            If True, do not include images with no annotations in the dataset
        """

        def __init__(
            self,
            kwcoco_dataset: kwcoco.CocoDataset,
            image_metadata: Dict[int, Dict[str, Any]],
            skip_no_anns: bool = False,
        ):
            self._kwcoco_dataset = kwcoco_dataset

            self._image_ids = list()
            self._annotations = dict()

            for _, img_id in enumerate(kwcoco_dataset.imgs.keys()):
                bboxes = np.empty((0, 4))
                labels = []
                scores = []

                if img_id in kwcoco_dataset.gid_to_aids.keys() and len(kwcoco_dataset.gid_to_aids[img_id]) > 0:
                    det_ids = kwcoco_dataset.gid_to_aids[img_id]
                    for det_id in det_ids:
                        ann = kwcoco_dataset.anns[det_id]

                        labels.append(ann["category_id"])

                        if "score" in ann:
                            scores.append(ann["score"])
                        elif "prob" in ann:
                            scores.append(max(ann["prob"]))
                        else:
                            scores.append(1.0)

                        x, y, w, h = ann["bbox"]
                        bbox = [x, y, x + w, y + h]
                        bboxes = np.vstack((bboxes, bbox))
                elif skip_no_anns:
                    continue

                img_file = kwcoco_dataset.get_image_fpath(img_id)
                if not img_file.exists():
                    continue
                self._image_ids.append(img_id)
                self._annotations[img_id] = JATICDetectionTarget(
                    boxes=bboxes, labels=np.asarray(labels), scores=np.asarray(scores)
                )

            self._image_metadata = copy.deepcopy(image_metadata)
            self._image_metadata = {
                image_id: image_md for image_id, image_md in self._image_metadata.items() if image_id in self._image_ids
            }
            if len(self._image_metadata) != len(self._image_ids):
                raise ValueError("Image metadata length mismatch, metadata needed for every image.")

        def __len__(self) -> int:
            """Returns the number of images in the dataset."""
            return len(self._image_ids)

        def __getitem__(self, index: int) -> OBJ_DETECTION_DATUM_T:
            """Returns the dataset object at the given index."""
            image_id = self._image_ids[index]
            img_file = self._kwcoco_dataset.get_image_fpath(image_id)
            image = Image.open(img_file)
            width, height = image.size

            gid_to_aids = self._kwcoco_dataset.gid_to_aids

            self._image_metadata[image_id].update(
                dict(
                    id=image_id,
                    image_info=dict(width=width, height=height, file_name=img_file),
                    det_ids=(list(gid_to_aids[image_id]) if image_id in gid_to_aids else list()),
                )
            )

            image_array = np.asarray(image)
            if image.mode == "L":
                image_array = np.expand_dims(image_array, axis=2)

            input_img, dets, metadata = (
                np.asarray(np.transpose(image_array, axes=(2, 0, 1))),
                self._annotations[image_id],
                self._image_metadata[image_id],
            )

            return input_img, dets, metadata


class JATICObjectDetectionDataset(Dataset):
    """Implementation of the JATIC Object Detection dataset wrapper for dataset images of varying sizes.

    Parameters
    ----------
    imgs : Sequence[np.ndarray]
        Sequence of images.
    dets : Sequence[ObjectDetectionTarget]
        Sequence of detections for each image.
    metadata : Sequence[Dict[str, Any]]
        Sequence of custom metadata for each image.
    """

    def __init__(
        self,
        imgs: Sequence[np.ndarray],
        dets: Sequence[TargetType],
        metadata: Sequence[DatumMetadataType],
    ) -> None:
        self.imgs = imgs
        self.dets = dets
        self.metadata = metadata

    def __len__(self) -> int:
        return len(self.imgs)

    def __getitem__(self, index: int) -> OBJ_DETECTION_DATUM_T:
        return self.imgs[index], self.dets[index], self.metadata[index]
