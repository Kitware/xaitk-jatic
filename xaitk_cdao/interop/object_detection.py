from typing import Dict, Hashable, Iterable, Sequence, Tuple
import numpy as np

from smqtk_detection import DetectImageObjects
from smqtk_image_io import AxisAlignedBoundingBox

import maite.protocols.object_detection as od


class JATICDetector(DetectImageObjects):
    """
    Adapter for JATIC object detection protocol.

    :param detector: The JATIC protocol-based detector.
    :param id2name: Mapping from label IDs to names.
    :param img_batch_size: Image batch size for inference.
    """

    def __init__(
        self,
        detector: od.Model,
        id2name: Dict[int, Hashable],
        img_batch_size: int = 1
    ):
        self._detector = detector
        self._id2name = dict(sorted(id2name.items()))
        self._img_batch_size = img_batch_size

    def detect_objects(
      self,
      img_iter: Iterable[np.ndarray]
    ) -> Iterable[Iterable[Tuple[AxisAlignedBoundingBox, Dict[Hashable, float]]]]:
        all_out = list()

        # Convert from channels last to channels first
        # Channels first is specified in protocols v0.5.0
        def _to_channels_first(img: np.ndarray) -> np.ndarray:
            if img.ndim < 3:
                return img
            return np.moveaxis(img, -1, 0)

        # Combine JATIC detections for same bbox into one DetectImageObject detection
        def _xform_dets(
            bboxes: Iterable[AxisAlignedBoundingBox],
            labels: Sequence[Hashable],
            probs: np.ndarray
        ) -> Iterable[Tuple[AxisAlignedBoundingBox, Dict[Hashable, float]]]:

            dets_dict: Dict[AxisAlignedBoundingBox, Dict[Hashable, float]] = dict()
            for box, label, prob in zip(bboxes, labels, probs):
                if box not in dets_dict:
                    dets_dict[box] = {la: 0. for la in self._id2name.values()}
                dets_dict[box][label] = prob

            return [(box, prob_dict) for box, prob_dict in dets_dict.items()]

        # Convert bboxes into AxisAlignedBoundingBoxes
        # Protocols specify X0, Y0, X1, Y1 format in v0.5.0
        def _xform_bbox(box: np.ndarray) -> AxisAlignedBoundingBox:
            return AxisAlignedBoundingBox(box[0:2], box[2:4])

        # Convert protocol-based detector output into DetectImageObject format
        def _generate_outputs(batch: Sequence[np.ndarray]) -> None:
            predictions = self._detector(np.asarray(batch))

            for pred in predictions:
                boxes = [_xform_bbox(box) for box in np.asarray(pred.boxes)]
                labels = [self._id2name[la] for la in np.asarray(pred.labels)]
                scores = np.asarray(pred.scores)

                all_out.append(
                    _xform_dets(bboxes=boxes, labels=labels, probs=scores)
                )

        # Batch model passes
        batch = list()
        for img in img_iter:
            batch.append(_to_channels_first(img))

            if len(batch) == self._img_batch_size:
                _generate_outputs(batch)
                batch = list()

        # Leftover batch
        if len(batch) > 0:
            _generate_outputs(batch)

        return all_out

    def get_config(self) -> dict:
        raise NotImplementedError(
            "Constructor arguments are not serializable as is and require"
            "further implementation to do so."
        )
