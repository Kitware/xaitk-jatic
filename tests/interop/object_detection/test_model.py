from collections.abc import Hashable, Iterable, Sequence
from contextlib import AbstractContextManager
from typing import Union
from unittest.mock import MagicMock

import maite.protocols.object_detection as od
import numpy as np
import pytest
from smqtk_core.configuration import configuration_test_helper
from smqtk_image_io import AxisAlignedBoundingBox

from xaitk_jatic.interop.object_detection.model import JATICDetector

rng = np.random.default_rng()


class TestJATICObjectDetector:
    dummy_id_to_name = {0: "A", 1: "B", 2: "C"}
    dummy_boxes = np.asarray([[1, 2, 3, 4], [1, 2, 3, 4], [5, 6, 7, 8]])
    dummy_scores = np.asarray([0.25, 0.75, 0.95])
    dummy_labels = np.asarray([0, 2, 0])
    dummy_out = MagicMock(
        spec=od.ObjectDetectionTarget,
        boxes=dummy_boxes,
        labels=dummy_labels,
        scores=dummy_scores,
    )
    dummy_expected = [
        [
            (AxisAlignedBoundingBox([1, 2], [3, 4]), {0: 0.25, 1: 0.0, 2: 0.75}),
            (AxisAlignedBoundingBox([5, 6], [7, 8]), {0: 0.95, 1: 0.0, 2: 0.0}),
        ],
    ]

    @pytest.mark.parametrize(
        ("detector", "id_to_name", "img_batch_size", "expectation"),
        [
            (
                MagicMock(spec=od.Model),
                dummy_id_to_name,
                2,
                pytest.raises(NotImplementedError, match=r"Constructor arg"),
            ),
        ],
    )
    def test_configuration(
        self,
        detector: od.Model,
        id_to_name: dict[int, Hashable],
        img_batch_size: int,
        expectation: AbstractContextManager,
    ) -> None:
        """Test configuration stability."""
        inst = JATICDetector(detector=detector, ids=id_to_name.keys(), img_batch_size=img_batch_size)
        with expectation:
            configuration_test_helper(inst)

    @pytest.mark.parametrize(
        ("detector_output", "id_to_name", "img_batch_size", "imgs", "expected_return"),
        [
            (
                [dummy_out],
                dummy_id_to_name,
                1,
                [rng.integers(0, 255, (3, 256, 256), dtype=np.uint8)],
                dummy_expected,
            ),
            (
                [dummy_out],
                dummy_id_to_name,
                1,
                [rng.integers(0, 255, (256, 256), dtype=np.uint8)],
                dummy_expected,
            ),
            (
                [dummy_out] * 2,
                dummy_id_to_name,
                2,
                rng.integers(0, 255, (2, 256, 256), dtype=np.uint8),
                [dummy_expected[0]] * 2,
            ),
            (
                [MagicMock(spec=od.ObjectDetectionTarget, boxes=[], labels=[], scores=[])],
                dummy_id_to_name,
                1,
                rng.integers(0, 255, (1, 256, 256), dtype=np.uint8),
                [[]],
            ),
        ],
        ids=["single 3 channel", "single greyscale", "multiple images", "no dets"],
    )
    def test_smoketest(
        self,
        detector_output: Sequence[od.ObjectDetectionTarget],
        id_to_name: dict[int, Hashable],
        img_batch_size: int,
        imgs: Union[np.ndarray, Sequence[np.ndarray]],
        expected_return: Iterable[Iterable[tuple[AxisAlignedBoundingBox, dict[Hashable, float]]]],
    ) -> None:
        """Test that MAITE detector output is transformed appropriately."""
        mock_detector = MagicMock(spec=od.Model, return_value=detector_output)

        inst = JATICDetector(detector=mock_detector, ids=id_to_name.keys(), img_batch_size=img_batch_size)
        res = list(inst.detect_objects(imgs))
        assert res == expected_return
