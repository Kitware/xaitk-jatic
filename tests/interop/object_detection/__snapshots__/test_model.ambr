# serializer version: 1
# name: TestJATICObjectDetector.test_smoketest[multiple images]
  list([
    list([
      tuple(
        <smqtk_image_io.bbox.AxisAlignedBoundingBox min_vertex=[1 2] max_vertex=[3 4]>,
        dict({
          0: 0.25,
          1: 0.0,
          2: 0.75,
        }),
      ),
      tuple(
        <smqtk_image_io.bbox.AxisAlignedBoundingBox min_vertex=[5 6] max_vertex=[7 8]>,
        dict({
          0: 0.95,
          1: 0.0,
          2: 0.0,
        }),
      ),
    ]),
    list([
      tuple(
        <smqtk_image_io.bbox.AxisAlignedBoundingBox min_vertex=[1 2] max_vertex=[3 4]>,
        dict({
          0: 0.25,
          1: 0.0,
          2: 0.75,
        }),
      ),
      tuple(
        <smqtk_image_io.bbox.AxisAlignedBoundingBox min_vertex=[5 6] max_vertex=[7 8]>,
        dict({
          0: 0.95,
          1: 0.0,
          2: 0.0,
        }),
      ),
    ]),
  ])
# ---
# name: TestJATICObjectDetector.test_smoketest[no dets]
  list([
    list([
    ]),
  ])
# ---
# name: TestJATICObjectDetector.test_smoketest[single 3 channel]
  list([
    list([
      tuple(
        <smqtk_image_io.bbox.AxisAlignedBoundingBox min_vertex=[1 2] max_vertex=[3 4]>,
        dict({
          0: 0.25,
          1: 0.0,
          2: 0.75,
        }),
      ),
      tuple(
        <smqtk_image_io.bbox.AxisAlignedBoundingBox min_vertex=[5 6] max_vertex=[7 8]>,
        dict({
          0: 0.95,
          1: 0.0,
          2: 0.0,
        }),
      ),
    ]),
  ])
# ---
# name: TestJATICObjectDetector.test_smoketest[single greyscale]
  list([
    list([
      tuple(
        <smqtk_image_io.bbox.AxisAlignedBoundingBox min_vertex=[1 2] max_vertex=[3 4]>,
        dict({
          0: 0.25,
          1: 0.0,
          2: 0.75,
        }),
      ),
      tuple(
        <smqtk_image_io.bbox.AxisAlignedBoundingBox min_vertex=[5 6] max_vertex=[7 8]>,
        dict({
          0: 0.95,
          1: 0.0,
          2: 0.0,
        }),
      ),
    ]),
  ])
# ---
